### ISSUE TITLE
Ammonite not working under fish shell

### ISSUE THREAD

**Issue by mdedetrich** [lihaoyi/Ammonite#813]
When running Ammonite under fish shell, you get the following

```
Failed to execute process '/usr/local/bin/amm'. Reason:
exec: Exec format error
The file '/usr/local/bin/amm' is marked as an executable but could not be run by the operating system.
```

The issue is that you don't specify under what shell to run Ammonite under, i.e. if I edit the shell script to start ammonite, I see

```bash
@ 2>/dev/null # 2>nul & echo off & goto BOF^M
:
exec java -Xmx500m -XX:+UseG1GC $JAVA_OPTS -cp "$0" ammonite.Main "$@"
....
```

It instead should have the shebang, i.e. something like

```bash
#!/bin/bash
@ 2>/dev/null # 2>nul & echo off & goto BOF^M
:
exec java -Xmx500m -XX:+UseG1GC $JAVA_OPTS -cp "$0" ammonite.Main "$@"
....
```

**Comment by mdedetrich** [comment]
Found the issue here, https://github.com/lihaoyi/mill/blob/master/main/src/mill/modules/Jvm.scala Going to submit a PR on mill

**Comment by lihaoyi** [comment]
@mdedetrich did you install Ammonite using the command below? It's meant to prepend the shebang for you:

```bash
sudo sh -c '(echo "#!/usr/bin/env sh" && curl -L https://github.com/lihaoyi/Ammonite/releases/download/1.1.2/2.12-1.1.2) > /usr/local/bin/amm && chmod +x /usr/local/bin/amm' && amm
```

Note that the command for the stable-version installation was incorrect until recently (yesterday?) and I just corrected it

**Comment by mdedetrich** [comment]
@lihaoyi I installed ammonite via homebrew, here is the formula

https://github.com/Homebrew/homebrew-core/blob/master/Formula/ammonite-repl.rb

Is it possible to update the formula?

**Comment by lihaoyi** [comment]
@mdedetrich sure send a PR to homebrew

**Comment by ilovezfs** [comment]
@lihaoyi Can you please provide a release download that includes the shebang line? That's not the kind of logic that belongs in the Homebrew formula. Thanks!

**Comment by mdedetrich** [comment]
@lihaoyi I agree with @ilovezfs , can you re-open the ticket and let me know when mill is republish. I can send a PR that properly adds the shebang as well as updating the documentation.

**Comment by lihaoyi** [comment]
@mdedetrich done; you'll probably need to make it fall back to uploading separate windows/bash artifacts in that case, since the goal of removing the shebang line was to try and unify the two scripts (https://github.com/lihaoyi/Ammonite/pull/789)

**Comment by lihaoyi** [comment]
FWIW whatever solution we come up for this will also probably need to be ported to the Mill launcher script, which shares the same problem

**Comment by LolHens** [comment]
What if we always appended the shebang line. This would cause an error message on windows.
I experimented a bit and came up with this to remove the error message after it is printed:
https://gist.github.com/LolHens/00ccf1d14363411e0917bc1a8fad67e6
This has to be configurable of course.
Windows will still print 'C:\Users\xxxxxxx\Documents>#!/bin/sh"' but at least there would be no error message that would not confuse people and it will make it much easier to install for linux people.

**Comment by robby-phd** [comment]
While I think it's ideal to have separate packages for different platforms, in this case, the only difference is a single shebang line for specific shells such as fish, ksh, etc.; so it is a matter of uploading tens of MBs for every commit of amm (and mill) with just a single line that differs. Granted, GitHub does not put a limit on release sizes now, but it seems like a waste to me.

I don't use homebrew so I'm not familiar with it, but perhaps a compromise would be to change the installation setup for homebrew+fish/ksh/etc. users to something like the following (someone can figure the a single line command for this, I'm sure):

1. install amm/mill from homebrew as usual
2. prepend shebang to amm/mill installed by homebrew

For example, the setup instruction for Cygwin provides a sed command to [modify the path to use cygdrive](https://github.com/lihaoyi/Ammonite/blob/master/readme/Sample.scala#L22), and it's used for [testing mill under Cygwin in AppVeyor](https://github.com/lihaoyi/mill/blob/master/.appveyor.yml#L46).

(This compromise works if homebrew does not detect integrity of installed files via hashing when running certain tasks and decides it should reinstall stuff.)

**Comment by mdedetrich** [comment]
@robby-phd In general, the homebrew guys aren't really happy with modifying files directly unless strictly necessary, their argument is that this should be fixed upstream (see https://github.com/Homebrew/homebrew-core/pull/28053#issuecomment-390472441 for more info). I would also comment in that thread if you want to convince the homebrew guys that you really don't want to make a separate distribution just for nix.

From what I know, if you wan't to be really strict about not making separate distributions, technically I would have to use homebrews `patch` [command](https://docs.brew.sh/Formula-Cookbook#patches) so it actually verifies the checksum of the resulting modified file

**Comment by ilovezfs** [comment]
My suggestion is to have an installation script in the zip file that takes a prefix parameter.

```
./install.sh --prefix=/somewhere
```
or
```
./install.sh /somewhere
```

and then it can do whatever magic is necessary, and we'll happily run

```
def install
  system "./install.sh", prefix
end
```
in the formula.

**Comment by ilovezfs** [comment]
For example https://github.com/Homebrew/homebrew-core/blob/master/Formula/clojure.rb#L12-L14

**Comment by lihaoyi** [comment]
Using `patch` or having an install script both sound good to me

**Comment by AesaKamar** [comment]
Also encountered this when I installed `ammonite` via `nix`

**Comment by tonylotts** [comment]
Why not have just two separate launchers? One for unix/linux, written in sh or bash, with the appropriate shebang line included.  One for windows, written in <whatever shell language is available on windows>?

**Comment by sake92** [comment]
@tonylotts I agree with you. That's what all other tools/apps like maven, gradle, tomcat do.  
Another issue with current solution is that shebang **has to be the first line**.  So there's no way to do a check for OS or a jump..  
Also, feels so unnatural to change the whole JAR file just to tweak a bit of launcher script..   
What about using ammonite in different environments? I have to rename it always, once it's `amm`, on win it's `amm.bat` etc.

**Comment by mdedetrich** [comment]
@patrick-premont If you do this then you would have to manually edit the launcher in ammonite to add the `#!/bin/bash`

Personally I just run `sh` before running ammonite, then you can install it as you wish

**Comment by swiesner-dlr** [comment]
I use a custom `amm` function in my Fish configuration to work around this issue:

```
function amm --description 'Scala REPL'
    sh -c 'amm "$@"' amm $argv
end
```

**Comment by patrickf3139** [comment]
Thanks @swiesner-dlr . Works for me!

**Comment by patrickf3139** [comment]
Looks like this issue is fixed now, as your PR has been merged and released, right @mdedetrich ?

**Comment by jsatk** [comment]
> I use a custom `amm` function in my Fish configuration to work around this issue:
> 
> ```
> function amm --description 'Scala REPL'
>     sh -c 'amm "$@"' amm $argv
> end
> ```

Is this still needed? Does ammonite work w/o this script in fish shell?

**Comment by laughedelic** [comment]
It seems to work for me well without any workarounds on Fish 3.2.2 and Ammonite 2.4.0.

**Comment by jsatk** [comment]
Excellent. Will delete my wrapper script. Thanks @laughedelic.

**Comment by jsatk** [comment]
@lihaoyi it appears this can be closed.
